### PR TITLE
Feature/ab#221 template documentation

### PR DESCRIPTION
**Pull request by witchpou**
Documentation of template usage.

### CODE REVIEW

**Comment by DerCed** [comment]
Link auf der Startseite zu dem Template geht nicht :(

File: docs/templates-howto.md [hunk 1]
```diff
@@ -1,130 +1,160 @@
...
-For creating a template, you need a project saved into a git repository. The generator with substitude all occurrences of the name in the apptemplate later on.
+- [How load existing Templates](#how-to-load-existing-templates)
+- [How to Create a Template](#how-to-create-an-own-template)
+
+# How to load Existing Templates
+## How to use the Default Template
+
+In order to start with templating, you can use an already existing template. A default react template can be found [here](https://github.com/starwit/reacthook-spring-template).
+
+1. If it is not already existing, you have to add it to the ProjectBuilder.
```
Comment (DerCed) [inline docs/templates-howto.md:19]:
```suggestion
1. If not already done, add your Template to ProjectBuilder.
```

File: docs/templates-howto.md [hunk 2]
```diff
@@ -1,130 +1,160 @@
...
+
+    * You can find the menu-item TEMPLATES in the header menu. It opens a list of all registered templates.
```
Comment (DerCed) [inline docs/templates-howto.md:21]:
```suggestion
    * You can find the menu-item `TEMPLATES` in the header menu. It opens a list of all registered templates.
```

File: docs/templates/template-def-v1.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+| Name        | Description     |      Example  |
+| ------------- |-------------| -------------|
+| projecthome   | home directory | `${projecthome}` |
+| project.targetPath | directory where the template is checked out  | `${projecthome}/${project.targetPath}/generator-templates/entity/entity.ftl` |
+| project.title | name of the project | `${projecthome}/${project.targetPath}/${project.title}/persistence/` |
+| project.template | project template used - not used in templates |  |
+| project.packagePrefix | you can choose a package name which will be included in the Java-package path | `${projecthome}/${project.targetPath}/${project.title}/persistence/src/main/java/de/${project.packagePrefix?lower_case}/${project.title?lower_case}/entity/`
+| project.description | description of the project which is not used in templates warning: not possible anymore | |
+| project.domains | domain objects.   | |
+
```
Comment (DerCed) [inline docs/templates/template-def-v1.md:26]: Formatting

File: docs/templates/template-def-v1.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+			<#if !attribute.nullable>
+```
+
+#### Properties
+
+| Name        | Description     |      Example  |
+| ------------|-------------| -------------|
+| domain.name | name of the domain | public class ${domain.name}Entity extends AbstractEntity { |
+| domain.description | description of the domain | normally not used |
+| domain.getAttributes() | used to get all attributes of a domain | <#list (domain.getAttributes()) as attribute> |
```
Comment (DerCed) [inline docs/templates/template-def-v1.md:53]: Formatting

File: docs/templates/template-def-v1.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+| Name        | Description     |      Example  |
+| ------------|-------------| -------------|
+| attribute.name | name of the attribute | @Column(name="${attribute.name?upper_case}" |
+| attribute.description | description of the attribute | normally not used |
+| attribute.pattern | Regex for validation of input | @Pattern(regexp = "${attribute.pattern}")|
+| attribute.dataType | | <#if attribute.dataType == "BigDecimal"> |
+| attribute.min | String: minimum length of the String. Number: smallest number | |
+| attribute.max | String: maximum length of the String. Number: highest number | |
+| attribute.nullable | Is false if attribut value is required | <#if !attribute.nullable>, nullable = false</#if> |
```
Comment (DerCed) [inline docs/templates/template-def-v1.md:86]: Formatting

File: docs/templates/template-def-v2.md [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+package de.${app.packageName?lower_case}.persistence.entity;
+
+<#list (imports) as import>
+${import}
+</#list>
+...
+```
+#### **Properties**
+
+| Name        | Type | Description     |
```
Comment (DerCed) [inline docs/templates/template-def-v2.md:31]: A little bit better but still formatting ;)

File: docs/templates/template-def-v2.md [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+| targetPath | String | directory where the template is checked out  |
+| baseName | String | name of the app |
+| template | AppTemplate | app template used - not used in templates |
+| packageName | String | you can choose a package name which will be included in the Java-package path |
+| entities | Array of EntityDto | entities   |
+| enums | Array of EnumDTo | enums   |
+
+### **entity**
+
+In Swagger: EntityDto
```
Comment (DerCed) [inline docs/templates/template-def-v2.md:42]: A link would be nice here.

File: docs/templates/template-def-v2.md [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+parent: Templates
+---
+
+# Template Definition Version 2
+
+The following objects can be used in template files. See also https://pb.starwit.de/ljprojectbuilder/swagger-ui.html
+
+### **app**
+
+In Swagger: AppDto
```
Comment (DerCed) [inline docs/templates/template-def-v2.md:14]: A link would be nice here

File: docs/templates/template-def-v2.md [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+  </#if>
+  </#list>
+</#if>
+```
+
+#### **Properties**
+
+| Name        | Type | Description     |
+| ------------|-------------| -------------|
+| otherEntityName | String | name of the entity on the other side |
```
Comment (DerCed) [inline docs/templates/template-def-v2.md:158]: Nitpicking here: The dataType would be nice to have in `code` blocks

File: docs/templates/template-def-v2.md [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+            "${enumItem?trim}"<#sep>,</#sep></#list>
+        ]
+    },
+...
+```
+#### **Properties**
+
+| Name        | Type | Description     |
+| ------------|-------------| -------------|
+| name | String | enum name |
```
Comment (DerCed) [inline docs/templates/template-def-v2.md:203]:
```suggestion
| name | String | name of the enum |
```

File: docs/templates/template-def-v1.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+		return ${attribute.name};
+	}
+</#if>
+```
+
+#### Properties
+
+| Name        | Description     |      Example  |
+| ------------|-------------| -------------|
+| attribute.name | name of the attribute | @Column(name="${attribute.name?upper_case}" |
```
Comment (DerCed) [inline docs/templates/template-def-v1.md:80]: Type is missing

File: docs/templates/template-def-v2.md [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+
+<#list (imports) as import>
+${import}
+</#list>
+...
+```
+#### **Properties**
+
+| Name        | Type | Description     |
+| ------------- |-------------| -------------|
```
Comment (DerCed) [inline docs/templates/template-def-v2.md:32]: An example column is missing

File: docs/templates/template-def-v2.md [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+  <#if field.fieldValidateRulesPattern?? && field.fieldValidateRulesPattern?length &gt; 0 && field.fieldType == "String">
+    @Pattern(regexp = "${field.fieldValidateRulesPattern}")
+  </#if>
+  <#if field.fieldType == "String">
+            <#if field.required>
+```
+
+#### **Properties**
+
+| Name        | Type | Description     |      Example  |
```
Comment (DerCed) [inline docs/templates/template-def-v2.md:78]: Those columns are perfect ;)

File: docs/templates/template-def-v2.md [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+            <#if field.required>
+    @NotBlank
+            </#if>
+            <#if field.fieldValidateRulesMinlength?? && field.fieldValidateRulesMaxlength??>
+    @Size(min = ${field.fieldValidateRulesMinlength}, max = ${field.fieldValidateRulesMaxlength})
+```
+
+#### **Properties**
+
+| Name        | Type | Description     |
```
Comment (DerCed) [inline docs/templates/template-def-v2.md:111]: Examples are missing

File: docs/templates/template-def-v2.md [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+  <#elseif relation.relationshipType == "OneToOne">
+    ...
+  </#if>
+  </#list>
+</#if>
+```
+
+#### **Properties**
+
+| Name        | Type | Description     |
```
Comment (DerCed) [inline docs/templates/template-def-v2.md:156]: Examples?

File: docs/templates/template-def-v2.md [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+        enumName: "${field.enumDef.name?trim?uncap_first}",
+        selectList: [<#list (field.enumDef.selectList) as enumItem>
+            "${enumItem?trim}"<#sep>,</#sep></#list>
+        ]
+    },
+...
+```
+#### **Properties**
+
+| Name        | Type | Description     |
```
Comment (DerCed) [inline docs/templates/template-def-v2.md:201]: Examples

File: docs/templates/template-def-v1.md [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+		<#if attribute.pattern?? && attribute.pattern?length &gt; 0 && attribute.dataType == "String">
+	@Pattern(regexp = "${attribute.pattern}")
+		</#if>
+		<#if attribute.dataType == "String">
+			<#if !attribute.nullable>
+```
+
+#### Properties
+
+| Name        | Description     |      Example  |
```
Comment (DerCed) [inline docs/templates/template-def-v1.md:49]: The Data-Type is missing here

File: docs/templates/template-def-v1.md [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+# Template Definition Version 1 <sup>depricated</sup>
+
+The following objects can be used in template files.
+
+### **project**
+
+General information about the project are stored.
+
+
+| Name        | Description     |      Example  |
```
Comment (DerCed) [inline docs/templates/template-def-v1.md:17]: Data Type column is missing

File: docs/templates-howto.md [hunk 19]
```diff
@@ -1,130 +1,160 @@
...
+|category | String | no | You can use it for getting a better overview of yout template configurations | `"category": "ENTITY"` |
 
-# How to Configure Apptemplates
+### Freemarker Substitutions for TemplatePath and TargetPath
 
-### **project**
+The following substitutions are valid in template path:
 
-The project-object contains general information and can be used to configure the paths for templatefiles-sets saved as apptemplate.
+* `app` (or `project`<sup>depricated</sup>): all properties of the app defined in ProjectBuilder. See AppDto definition on https://pb.starwit.de/ljprojectbuilder/swagger-ui.html or definitions below
```
Comment (DerCed) [inline docs/templates-howto.md:125]: I would generally remove the deprecated parts of this documentation. The user should build templates with the not deprecated parts...

```suggestion
* `app` (or `project`<sup>deprecated</sup>): all properties of the app defined in ProjectBuilder. See AppDto definition on https://pb.starwit.de/ljprojectbuilder/swagger-ui.html or definitions below
```

**Review by DerCed — CHANGES_REQUESTED** [review]
Finde die Dokumentation sehr gelungen! Habe noch ein paar Anmerkungen angebracht, die das noch ein wenig vervollständigen könnten

File: docs/templates/template-def-v1.md (hunk 3 above)
Comment (witchpou) [reply docs/templates/template-def-v1.md:26]: > 

was meinst du mit Formatting? Was soll ich ändern?

File: docs/templates/template-def-v2.md (hunk 15 above)
Comment (witchpou) [reply docs/templates/template-def-v2.md:156]: are placed above

File: docs/templates/template-def-v1.md (hunk 17 above)
Comment (witchpou) [reply docs/templates/template-def-v1.md:49]: are placed above

Comment (witchpou) [reply docs/templates/template-def-v1.md:49]: it's deprecated - I just want to let it as it was before

File: docs/templates/template-def-v1.md (hunk 18 above)
Comment (witchpou) [reply docs/templates/template-def-v1.md:17]: it's deprecated - I just want to let it as it was before

File: docs/templates/template-def-v2.md (hunk 16 above)
Comment (witchpou) [reply docs/templates/template-def-v2.md:201]: are placed above

File: docs/templates/template-def-v2.md (hunk 8 above)
Comment (witchpou) [reply docs/templates/template-def-v2.md:14]: good idea

**Pull request merged**